New FES Contributor Docs + FES Survey Link

almchung opened the pull request:
This PR contains the final drafts for the new FES contributor docs + updated FES inline docs. 

Changes:
- `friendly_error_system.md` doc is now splitted into two parts: `friendly_error_system.md` + `fes_reference_dev_notes.md`.
- `friendly_error_system.md`: completely re-written with a focus on the error message writing and the internationalization effort.
- `friendly_error_system.md`: link for [FES 2021 Survey](https://bit.ly/p5fesSurvey) added.
- `fes_reference_dev_notes.md`: dev notes + technical details. The "Thoughts for Future Work" section will be updated in the next round, based on results from the FES survey (~early 2022).
- `core/friendly_errors`: inline docs are re-organized. Some functions' inline docs now describe their connection to the translation files. Helpful but long inline comments were moved to `fes_reference_dev_notes.md`.

#### PR Checklist
- [X] `npm run lint` passes
- [X] [Inline documentation] is included / updated
- [ ] [Unit tests] are included / updated

almchung commented:
@outofambit @Qianqianye: These files are ready to be reviewed! I will also let my mentors know about this update. 
Thank you so much! 🙏

kjhollen commented on contributor_docs/fes_reference_dev_notes.md:
```diff
@@ -0,0 +1,456 @@
+  //PI is a p5.js reserved constant
+  let PI = 100;
+}
+// FES will show:
```
These are great examples. I've noticed here & elsewhere that this messages are long enough to cause a horizontal scroll when displayed in GitHub in the README. I don't know if there's anyway we can wrap the text instead, but it would be nice if it did. (Feel free to ignore this comment if it adds too much work)

almchung replied:
This is a good suggestion. I will pull out the FES output preview messages from the code blocks and put them into quotes instead. Quote should support word-wrap by default.

kjhollen commented on contributor_docs/fes_reference_dev_notes.md:
`````diff
@@ -0,0 +1,456 @@
+// FES will generate the following message in the console:
+// > 🌸 p5.js says: arc() was expecting Number for parameter #0 (zero-based index), received string instead. [http://p5js.org/reference/#p5/arc]
+````
+##### Description
`````
should description go before the examples here? I'm realizing that the pattern here matches the p5.js reference (leading with examples first), but for some of the functions (in particular fesCodeReader()) I'm not really sure what they do just from the function name.

kjhollen commented on contributor_docs/fes_reference_dev_notes.md:
```diff
@@ -0,0 +1,456 @@
+
+### `_friendlyAutoplayError()`
+##### Description
+`_friendlyAutoplayError()` is called internally if there is an error with autoplay.
```
I don't know what autoplay is 😅 maybe a short description of what this refers to would be helpful?

almchung replied:
Good catch! Will add a description here 😄

kjhollen commented on contributor_docs/fes_reference_dev_notes.md:
```diff
@@ -0,0 +1,456 @@
+
+* The extracting variable/function names feature of FES's `sketch_reader` is not perfect and some cases might go undetected (for eg: when all the code is written in a single line).
+
+## Thoughts for Future Work
```
great to have all of this documented!

almchung replied:
I understood this line as "extracting variable/function names from the user code" the feature may miss some cases due to coding style. I will re-write this so it will be clearer.

almchung replied:
I will add an item to the Thoughts for Future Work section about identifying these cases and update the reference.

kjhollen commented on src/core/friendly_errors/fes_core.js:
```diff
@@ -1,79 +1,6 @@
  * @for p5
  * @requires core
- *
- * This is the main file for the Friendly Error System (FES). Here is a
```
is the intent that markdown file for the contributor docs replaces this? might be nice to link from here somehow, if possible?

almchung replied:
I thought (1) this information would be easier to read with some markdown formatting and (2) the `fes_core.js` file was getting quite long and would benefit from modularization/organization (maybe for future work!). 
I agree this will be a great place to add a link to the fes_reference_dev_notes.md. 👍

kjhollen commented on src/core/friendly_errors/fes_core.js:
```diff
@@ -242,16 +170,17 @@ if (typeof IS_MINIFIED !== 'undefined') {
    */
-  p5._friendlyError = function(message, method, color) {
-    p5._report(message, method, color);
+  p5._friendlyError = function(message, func, color) {
```
I am wondering if the Contributor docs are generated/updated from this automatically? if not, might be a good work item to add to the "future work" section of the markdown file

kjhollen reviewed (commented):
Hi Alice, this looks like great progress! The main thing I'd want to see addressed before a merge is having an introduction or short technical overview of the FES functions in the contributor docs. Most other things I've noted are just questions that I have or suggestions for future work.

almchung commented on contributor_docs/fes_reference_dev_notes.md:
````diff
@@ -0,0 +1,456 @@
+```
+@param  {String}        message   Message to be printed
+@param  {String}        [func]    Name of function
+@param  {Number|String} [color]   CSS color string or error type
````
This should be just "CSS color code". Will change this!

almchung commented:
Updated reflecting feedback from @kjhollen. Specifically:
FES Reference + Dev Notes: 
1. Added an introduction paragraph 
2. Swapped orders of Description and Examples sections
3. Reformatted FES output examples to have auto word-wrap
4. Fixed typos such as parameter `CSS color string` for `_report()` and `_friendlyError()`
5. Expanded on `autoplay`
6. Added items on the Future Work section: `sketch_reader()` and generative contributor doc

almchung commented:
@Qianqianye @outofambit Planning to merge this by the end of the day!
